### ISSUE TITLE
Pipelines analytics

### ISSUE THREAD

**Issue by spacedub** [dagger/dagger#3788]
### What are you trying to do?

One of the challenges with CI systems right now is the lack (or poor state) of analytics.

As a requirement for us to use dagger as a building block, we would love to have access to execution data so to build solid "pipeline observability", as a core feature to our product.

### Why is this important to you?

Well, customers want it :)

Information about pipelines and "actions" execution time, failure trends, caching, are important to have for people in charge of maintaining and improving these pipelines.
To improve something, you need information about where is the problem, and be able to  measure the impact of your  changes.

Such info could also be used for metering, or otherwise building high-level views of pipelines performance across an organization.

### How are you currently working around this?

Our own implementation is currently built directly on top of buildkit which does provide the info we need.

You can check what `buildctl` does with the command line argument named `--trace` (*not* to be confused with actual tracing, unrelated).

For every vertex, we get:
- the name/description of the "action" (pretty much `metadata["llb.customname"]`)
- the progress group it was a part of
- the start and end times
- whether it was cached
- if it errored
- stdout and stderr logs
- digest of the action
- digests of inputs vertices

We use that ^ to rebuild the graph of what happened, and, enriched with some metadata about the pipeline, build dashboards and fancy visuals ("weighted" graphs, trends over time, etc).

"Where" that data is being sent / how visualization works is irrelevant here and IMHO should be kept as a separate, orthogonal problem (we are using ELK for now, but it should be easy to send that somewhere else). 

Obviously, a better, more refined data model than what buildkit is providing could be devised - would be happy to help with that.

cc @shykes since we just had that conversation over almond milk

cc @aluzzardi since #2140 is talking about "High-level summaries/analytics about cache use, built on top of observability features. I.e. the ability to collect statistics about cache performance across many runs."

self-cc @dubo-dubon-duponey

**Comment by mgood** [comment]
Yes! I started to look into what's supported since I saw there is some tracing-related code, but I didn't see any documentation for it yet.

Something I'd love to see that I don't see mentioned here is a way to pass the tracing context all the way down so that I could add my own tracing to the jobs being run inside Dagger. For example, on a test run I'd love to have the ability to trace execution within the individual tests. This would be invaluable in pinpointing testing bottlenecks, or sources of flakey test failures.
